sunscreen_math/src/metal_impl/ristrettovec.rs

```diff
@@ -0,0 +1,716 @@
+    fn buffer_slice_mut(&mut self) -> &mut [u32] {
+        let byte_len = self.len * size_of::<RistrettoPoint>();
+
+        unsafe { slice::from_raw_parts_mut(self.data.contents() as *mut u32, byte_len) }
+    }
+
+    fn buffer_slice(&self) -> &[u32] {
+        let byte_len = self.len * size_of::<RistrettoPoint>();
+
+        unsafe { slice::from_raw_parts(self.data.contents() as *const u32, byte_len) }
+    }
+}
+
+impl Add<RistrettoPointVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn add(self, rhs: RistrettoPointVec) -> Self::Output {
+        &self + &rhs
+    }
+}
+
+impl Add<&RistrettoPointVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn add(self, rhs: &RistrettoPointVec) -> Self::Output {
+        &self + rhs
+    }
+}
+
+impl Add<RistrettoPointVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn add(self, rhs: RistrettoPointVec) -> Self::Output {
+        self + &rhs
+    }
+}
+
+impl Add<&RistrettoPointVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn add(self, rhs: &RistrettoPointVec) -> Self::Output {
+        assert_eq!(self.len(), rhs.len());
+
+        let runtime = Runtime::get();
+
+        let len = self.len_bytes();
+
+        let o = Self::Output {
+            data: runtime.alloc(len),
+            len: rhs.len(),
+        };
+
+        let len_gpu = U32Arg::new(rhs.len() as u32);
+
+        // TODO: o gets mutated here. Need to figure out what that means in terms of UB.
+        runtime.run(
+            "ristretto_add",
+            &[&self.data, &rhs.data, &o.data, &len_gpu.data],
+            [(rhs.len() as u64, 64), (1, 1), (1, 1)],
+        );
+
+        o
+    }
+}
+
+impl Sub<RistrettoPointVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn sub(self, rhs: RistrettoPointVec) -> Self::Output {
+        &self - &rhs
+    }
+}
+
+impl Sub<&RistrettoPointVec> for RistrettoPointVec {
+    type Output = Self;
+
+    fn sub(self, rhs: &RistrettoPointVec) -> Self::Output {
+        &self - rhs
+    }
+}
+
+impl Sub<RistrettoPointVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn sub(self, rhs: RistrettoPointVec) -> Self::Output {
+        self - &rhs
+    }
+}
+
+impl Sub<&RistrettoPointVec> for &RistrettoPointVec {
+    type Output = RistrettoPointVec;
+
+    fn sub(self, rhs: &RistrettoPointVec) -> Self::Output {
+        assert_eq!(self.len(), rhs.len());
+
+        let runtime = Runtime::get();
+
+        let len = self.len_bytes();
+
+        let o = Self::Output {
+            data: runtime.alloc(len),
+            len: rhs.len(),
+        };
+
+        let len_gpu = U32Arg::new(rhs.len() as u32);
+
+        // TODO: o gets mutated here. Need to figure out what that means in terms of UB.
+        runtime.run(
+            "ristretto_sub",
+            &[&self.data, &rhs.data, &o.data, &len_gpu.data],
+            [(rhs.len() as u64, 64), (1, 1), (1, 1)],
+        );
+
+        o
+    }
+}
```

Comment:
Can this at all be simplified by calling `add` but negating the second argument? Does that add overhead?

Reply:
Curious about this too.. my guess is that the operations are specific to unsigned arithmetic?

Reply:
It would be a bit of overhead. You would have to allocate a temporary *vec* of elements to store the negation and run 2 GPU kernels instead of 1.